### ISSUE TITLE
[System.Linq.Expressions] Allow Switch with empty cases list

### ISSUE THREAD

**Issue by ViIvanov** [dotnet/corefx#1877]
Now we can not create an expression, that represents this code:
```cs
switch(x) {
default:
    return 0;
}
```
But c# allows this and this is can be very helpful - users do not needs a checks and  brunches in own code.

Problem with a next line: https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Linq/Expressions/SwitchExpression.cs#L205

I have asked this question on StackOverflow a few months ago:  http://stackoverflow.com/questions/28360782/switch-without-cases-but-with-default-in-system-linq-expressions

Why this is not allowed? Can we change it?

**Comment by VSadov** [comment]
While it may seem trivial, it could be a risky change because it would require ensuring that all the consumers of such trees can handle switches without cases.

At this point, if situation is very common in some kind of code, I think an easier approach is to add a helper method, but based on whether case list is empty or not would construct a switch or would just return the default expression.
This might still be fairly convenient without a danger of breaking something not prepared to handle switches without cases.

**Comment by hackcraft** [comment]
What about something that, if the type of `x` is `T` takes an expression equivalent to:

    switch(x)
    {
      default:
        someMethodBody();
       break;
    }

and returns an expression equivalent to:

    switch(x)
    {
      default(T):
       someMethodBody();
       break;
      default:
        someMethodBody();
       break;
    }

Or perhaps better would be:

    switch(x)
    {
      default(T):
       goto default;
      default:
        someMethodBody();
       break;
    }

Then anything depending on the current rule that a switch cannot be just a default wouldn't break.

**Comment by hackcraft** [comment]
There's a stab at that. A helper method such as @VSadov suggests would still produce a tighter expression, but this should (if I haven't missed anything) mean that such switch expressions "just work".

**Comment by VSadov** [comment]
I will need to check with original designers of expressions API, but I think it is very likely that not supporting switch with no cases is intentional.

As I remember at some pre-release point expression trees supported a lot more nodes. There were for/foreach/while loops, ifs with multiple elseifs and such. Many high level constructs were too unwieldy to be created in one call, and instead required using builders that accumulate pieces and validate them upon realization of the construct. Some of these constructs had very complicated (and also buggy) wellformedness requirements.
The number of supported scenarios was getting bigger and bigger and even documenting all the scenarios was becoming a problem.

At some point it was decided that "reduced instruction set" is a more manageable approach for the core library both from implementation and from usability/documenting prospective.

As a result the core library was reduced (within reason) to have only minimum set of primitives that are hard/impossible to express in terms of other nodes. Any additional "convenience" helpers would be responsibility of libraries higher on the stack.

One of the results of such approach is orthogonality - there is generally just a single way to do the same thing. Since there is support for Expression.Empty, there is no support for empty blocks. 
As an added bonus there is no need to figure and document what happens if a block tries to have locals or a nonvoid type, but provides no expressions. 
Basically there is only one recommended way to have a "Nop". If there is a need to generate empty/nonempty blocks of expressions dynamically - just introduce a layer of instruction choosing helpers that can switch between Block and Empty. 

I will check, but my strong guess is that not supporting switch without cases is a direct result of this approach.

That would basically make this issue ByDesign and an instruction switcher on the outer layer would be an intended way of dealing with this situation.

**Comment by hackcraft** [comment]
It certainly would. And certainly I'm better off that TIL `Expression.Empty()` than trying to use an empty block. If it is ByDesign, I couldn't see anything about it in the documentation, so perhaps that could be fixed.
Though also if it is ByDesign, I'd make the argument for a change (but this would now be arguably an API change) on the basis that if you have code producing multiple numbers of cases in different situations, you have to special-case there being zero such cases produced.

**Comment by VSadov** [comment]
I think one of the expectations was that every heavy user of ET would inevitably have another layer of abstractions that are higher level and more expressive. It was somewhat expected that if there is enough need and commonality, then perhaps even a common library would emerge that would be used by compilers that target ETs.
I believe IronPython/IRonRuby have such layers and perhaps even share the implementation, but evidently there is no common "higher level" API to expression trees. If there was such layer, it would be the perfect place to put enhanced factories, builders, deduplicating logic and other value-add functionality. 
I guess there was not enough enough need (or commonality) to drive such library into existence. Or perhaps there is one and I am just not aware...

**Comment by hackcraft** [comment]
A higher-level constructor would do a better job; A method that takes zero-or-more cases and a default expression and returns either a `SwitchExpression` or that default would produce a tighter expression in the no-case condition, with a greater chance of it being reduced further nicely.

That said, turning a zero-case switch into a single-case switch does allow us to go from an error condition to something that "just works". (Assuming of course, that my suggestion does indeed "just work" and I'm not missing something).
